components/tidb_query_datatype/src/codec/mysql/duration.rs

```diff
@@ -204,20 +204,46 @@ mod parser {
         Ok((rest, frac * TEN_POW[NANO_WIDTH.saturating_sub(len)]))
     }
 
+    /// a string can match datetime format only if it starts with a series of digits
+    /// whose length matches the full format of DateTime literal (12, 14)
+    /// or the string start with a date literal
+    fn format_can_match_datetime(input: &str) -> IResult<(), (), ()> {
+        let (rest, digits) = digit1(input)?;
+
+        if digits.len() == 12 || digits.len() == 14 {
+            return Ok(((), ()));
+        }
+
+        let (rest, _) = anysep(rest)?;
+        let (rest, _) = digit1(rest)?;
+        let (rest, _) = anysep(rest)?;
+        let (rest, _) = digit1(rest)?;
+
+        if matches!(rest.chars().next(), Some(c) if c == 'T' || c == ' ') {
+            Ok(((), ()))
+        } else {
+            Err(nom::Err::Error(()))
+        }
+    }
+
     pub fn parse(
         ctx: &mut EvalContext,
         input: &str,
         fsp: u8,
-        fallback_to_daytime: bool,
+        fallback_to_datetime: bool,
         overflow_as_null: bool,
     ) -> Option<Duration> {
         let input = input.trim();
         if input.is_empty() {
-            return Some(Duration::zero());
+            return None;
         }
 
         let (rest, neg) = negative(input).ok()?;
         let (rest, _) = space0::<_, ()>(rest).ok()?;
+        let fallback_to_datetime = format_can_match_datetime(rest).is_ok() && fallback_to_datetime;
```

Comment:
sure, I will use a new variable instead to represent the return value of `format_can_match_datetime ` in order to avoid introducing `mut` modifier for the argument.

Reply:
If `fallback_to_datetime` is false, it's a waste to run `format_can_match_datetime`. 
What's the disadvantage to add a `mut` for `fallback_to_datetime`?

Reply:
I think bitwise & for boolean is not short-circuit operator.

Reply:
Indeed. So it should be `let fallback_to_datetime = fallback_to_datetime && format_can_match_datetime(rest).is_ok();`.

Reply:
fine